controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -666,6 +672,49 @@
         return store.deletePath(waitingRequestProcessorPath, false);
     }
 
+    @Override
+    CompletableFuture<Void> createWriterMarkRecord(String writer, long timestamp, ImmutableMap<Long, Long> position) {
+        String writerPath = getWriterPath(writer);
+        WriterMark mark = new WriterMark(timestamp, position);
+        return Futures.toVoid(store.createZNode(writerPath, mark.toBytes()));
+    }
+
+    @Override
+    public CompletableFuture<Void> removeWriter(String writer) {
+        String writerPath = getWriterPath(writer);
+        return store.deletePath(writerPath, false);
+    }
+
+    @Override
+    CompletableFuture<VersionedMetadata<WriterMark>> getWriterMarkRecord(String writer) {
+        String writerPath = getWriterPath(writer);
+        return store.getData(writerPath, WriterMark::fromBytes);
+    }
+
+    @Override
+    CompletableFuture<Void> updateWriterMarkRecord(String writer, long timestamp, ImmutableMap<Long, Long> position, Version version) {
+        String writerPath = getWriterPath(writer);
+        WriterMark mark = new WriterMark(timestamp, position);
+
+        return Futures.toVoid(store.setData(writerPath, mark.toBytes(), version));
```

Comment:
It seems like this would expensive. This is going to get called a LOT.

Reply:
Yes, but its inevitable. We need to store marks as these requests could be received on different controller instances. Besides we dont hold all of this in memory anyway.

Reply:
No, because writers call can reach any arbitrary controller. So either we build some mechanism with which all writers are redirected to the same controller instance to report marks for a given stream. 
Or we durably store the writer marks and then perform some computation on them asynchronously with some periodicity.
We decided to do the latter in the PDP. 
The disadvantage of doing former is, we need to make explicit grpc call into a specific controller instance or redirect the request from one controller to another. The controller will need to maintain in memory state for all writers, which in turn will put memory pressure on it. And in case of failure, we lose marks reported by all writers and that can lead to missing writer positions in the watermark computation if the writer doesnt report another mark.